File: VB/DevExpressMvcApplication1/Models/SchedulerDataHelper.vb
Imports Microsoft.VisualBasic
Imports System.Collections
Imports System.Linq
Imports DevExpressMvcApplication1
Imports DevExpress.Web.Mvc
Imports DevExpressMvcApplication1.Models
Imports System.Collections.Generic
Imports System.Drawing
Imports System
Imports System.Web
Imports System.Web.Mvc.Html
Imports DevExpress.XtraScheduler
Imports DevExpress.Web

Public Module SchedulerDataHelper
	Public Function GetResources() As List(Of CustomResource)
		Dim resources As New List(Of CustomResource)()
		resources.Add(CustomResource.CreateCustomResource(1, "Max Fowler", Color.Yellow.ToArgb()))
		resources.Add(CustomResource.CreateCustomResource(2, "Nancy Drewmore", Color.Green.ToArgb()))
		resources.Add(CustomResource.CreateCustomResource(3, "Pak Jang", Color.LightPink.ToArgb()))
		Return resources
	End Function

	Private myRand As New Random()
	Public Function GetAppointments(ByVal resources As List(Of CustomResource)) As List(Of CustomAppointment)
		Dim appointments As New List(Of CustomAppointment)()
		For Each item As CustomResource In resources
			Dim subjPrefix As String = item.Name & "'s "
			appointments.Add(CustomAppointment.CreateCustomAppointment(subjPrefix & "meeting", item.ResID, 2, 5, lastInsertedID))
			lastInsertedID += 1
			appointments.Add(CustomAppointment.CreateCustomAppointment(subjPrefix & "travel", item.ResID, 3, 6, lastInsertedID))
			lastInsertedID += 1
			appointments.Add(CustomAppointment.CreateCustomAppointment(subjPrefix & "phone call", item.ResID, 0, 10, lastInsertedID))
			lastInsertedID += 1
		Next item
		Return appointments
	End Function


	Public ReadOnly Property DataObject() As SchedulerDataObject
		Get
'INSTANT VB NOTE: The local variable dataObject was renamed since Visual Basic will not allow local variables with the same name as their enclosing function or property:
			Dim dataObject_Renamed As New SchedulerDataObject()


			If HttpContext.Current.Session("ResourcesList") Is Nothing Then
				HttpContext.Current.Session("ResourcesList") = GetResources()
			End If
			dataObject_Renamed.Resources = TryCast(HttpContext.Current.Session("ResourcesList"), List(Of CustomResource))

			If HttpContext.Current.Session("AppointmentsList") Is Nothing Then
				HttpContext.Current.Session("AppointmentsList") = GetAppointments(dataObject_Renamed.Resources)
			End If
			dataObject_Renamed.Appointments = TryCast(HttpContext.Current.Session("AppointmentsList"), List(Of CustomAppointment))
			Return dataObject_Renamed
		End Get
	End Property

	Private defaultAppointmentStorage_Renamed As MVCxAppointmentStorage
	Public ReadOnly Property DefaultAppointmentStorage() As MVCxAppointmentStorage
		Get
			If defaultAppointmentStorage_Renamed Is Nothing Then
				defaultAppointmentStorage_Renamed = CreateDefaultAppointmentStorage()
			End If
			Return defaultAppointmentStorage_Renamed

		End Get
	End Property

	Private Function CreateDefaultAppointmentStorage() As MVCxAppointmentStorage
		Dim appointmentStorage As New MVCxAppointmentStorage()
		appointmentStorage.AutoRetrieveId = True
		appointmentStorage.Mappings.AppointmentId = "ID"
		appointmentStorage.Mappings.Start = "StartTime"
		appointmentStorage.Mappings.End = "EndTime"
		appointmentStorage.Mappings.Subject = "Subject"
		appointmentStorage.Mappings.AllDay = "AllDay"
		appointmentStorage.Mappings.Description = "Description"
		appointmentStorage.Mappings.Label = "Label"
		appointmentStorage.Mappings.Location = "Location"
		appointmentStorage.Mappings.RecurrenceInfo = "RecurrenceInfo"
		appointmentStorage.Mappings.ReminderInfo = "ReminderInfo"
		appointmentStorage.Mappings.ResourceId = "OwnerId"
		appointmentStorage.Mappings.Status = "Status"
		appointmentStorage.Mappings.Type = "EventType"

		appointmentStorage.CustomFieldMappings.Add(New DevExpress.Web.ASPxScheduler.ASPxAppointmentCustomFieldMapping("AppointmentCustomField", "CustomInfo"))
		Return appointmentStorage
	End Function

	Private defaultResourceStorage_Renamed As MVCxResourceStorage
	Public ReadOnly Property DefaultResourceStorage() As MVCxResourceStorage
		Get
			If defaultResourceStorage_Renamed Is Nothing Then
				defaultResourceStorage_Renamed = CreateDefaultResourceStorage()
			End If
			Return defaultResourceStorage_Renamed

		End Get
	End Property

	Private Function CreateDefaultResourceStorage() As MVCxResourceStorage
		Dim resourceStorage As New MVCxResourceStorage()
		resourceStorage.Mappings.ResourceId = "ResID"
		resourceStorage.Mappings.Caption = "Name"
		resourceStorage.Mappings.Color = "Color"
		Return resourceStorage
	End Function

	Public Function GetSchedulerSettings() As SchedulerSettings
		Return GetSchedulerSettings(Nothing)
	End Function

	<System.Runtime.CompilerServices.Extension> _
	Public Function GetSchedulerSettings(ByVal customHtml As System.Web.Mvc.HtmlHelper) As SchedulerSettings
		Dim settings As New SchedulerSettings()
		settings.Name = "scheduler"
		settings.CallbackRouteValues = New With {Key .Controller = "Home", Key .Action = "SchedulerPartial"}
		settings.EditAppointmentRouteValues = New With {Key .Controller = "Home", Key .Action = "EditAppointment"}
		settings.CustomActionRouteValues = New With {Key .Controller = "Home", Key .Action = "CustomCallBackAction"}

		settings.Storage.Appointments.Assign(SchedulerDataHelper.DefaultAppointmentStorage)
		settings.Storage.Resources.Assign(SchedulerDataHelper.DefaultResourceStorage)

		settings.Storage.EnableReminders = False
		settings.GroupType = SchedulerGroupType.Resource
		settings.Views.DayView.Styles.ScrollAreaHeight = 400
		settings.Start = DateTime.Now

		settings.AppointmentFormShowing = Function(sender, e) AnonymousMethod1(sender, e)

			'ViewBag.ReminderDataSource = container.ReminderDataSource;
        settings.OptionsForms.SetAppointmentFormTemplateContent(Function(c) AnonymousMethod2(c, customHtml))
		Return settings
	End Function
	
	Private Function AnonymousMethod1(ByVal sender As Object, ByVal e As Object) As Boolean
		Dim scheduler = TryCast(sender, MVCxScheduler)
		If scheduler IsNot Nothing Then
			e.Container = New CustomAppointmentTemplateContainer(scheduler)
		End If
		Return True
	End Function
	
    Private Function AnonymousMethod2(ByVal c As Object, ByVal customHtml As System.Web.Mvc.HtmlHelper) As Boolean
        Dim container = CType(c, CustomAppointmentTemplateContainer)
        Dim modelAppointment As New ModelAppointment() With {.ID = If(container.Appointment.Id Is Nothing, -1, CInt(Fix(container.Appointment.Id))), .Subject = container.Appointment.Subject, .Location = container.Appointment.Location, .StartTime = container.Appointment.Start, .EndTime = container.Appointment.End, .AllDay = container.Appointment.AllDay, .Description = container.Appointment.Description, .EventType = CInt(Fix(container.Appointment.Type)), .Status = container.Appointment.StatusId, .Label = container.Appointment.LabelId, .CustomInfo = container.CustomInfo, .OwnerId = Convert.ToInt32(container.Appointment.ResourceId)}
        customHtml.ViewBag.DeleteButtonEnabled = container.CanDeleteAppointment
        TryCast(container.ResourceDataSource, ListEditItemCollection).RemoveAt(0)
        customHtml.ViewBag.ResourceDataSource = container.ResourceDataSource
        customHtml.ViewBag.StatusDataSource = container.StatusDataSource
        customHtml.ViewBag.LabelDataSource = container.LabelDataSource
        customHtml.RenderPartial("CustomAppointmentFormPartial", modelAppointment)
        Return True
    End Function

	Private lastInsertedID As Integer = 0

	' CRUD operations implementation
	Public Sub InsertAppointments(ByVal appts() As CustomAppointment)
		If appts.Length = 0 Then
			Return
		End If

		Dim appointmnets As List(Of CustomAppointment) = TryCast(HttpContext.Current.Session("AppointmentsList"), List(Of CustomAppointment))
		For i As Integer = 0 To appts.Length - 1
			If appts(i) IsNot Nothing Then
				appts(i).ID = lastInsertedID
				lastInsertedID += 1
				appointmnets.Add(appts(i))
			End If
		Next i
	End Sub

	Public Sub UpdateAppointments(ByVal appts() As CustomAppointment)
		If appts.Length = 0 Then
			Return
		End If

		Dim appointmnets As List(Of CustomAppointment) = TryCast(System.Web.HttpContext.Current.Session("AppointmentsList"), List(Of CustomAppointment))
        For i As Integer = 0 To appts.Length - 1
            Dim currentIndex As Integer = i
            Dim sourceObject As CustomAppointment = appointmnets.First(Function(apt) apt.ID = appts(currentIndex).ID)
            appts(i).ID = sourceObject.ID
            appointmnets.Remove(sourceObject)
            appointmnets.Add(appts(i))
        Next i
	End Sub

	Public Sub RemoveAppointments(ByVal appts() As CustomAppointment)
		If appts.Length = 0 Then
			Return
		End If

		Dim appointmnets As List(Of CustomAppointment) = TryCast(HttpContext.Current.Session("AppointmentsList"), List(Of CustomAppointment))
		For i As Integer = 0 To appts.Length - 1
            Dim currentIndex As Integer = i
            Dim sourceObject As CustomAppointment = appointmnets.First(Function(apt) apt.ID = appts(currentIndex).ID)
            appointmnets.Remove(sourceObject)
		Next i
	End Sub
End Module

File: CS/DevExpressMvcApplication1/Models/SchedulerDataHelper.cs
using System.Collections;
using System.Linq;
using DevExpressMvcApplication1;
using DevExpress.Web.Mvc;
using DevExpressMvcApplication1.Models;
using System.Collections.Generic;
using System.Drawing;
using System;
using System.Web;
using System.Web.Mvc.Html;
using DevExpress.XtraScheduler;
using DevExpress.Web;

public static class SchedulerDataHelper {
    public static List<CustomResource> GetResources() {
        List<CustomResource> resources = new List<CustomResource>();
        resources.Add(CustomResource.CreateCustomResource(1, "Max Fowler", Color.Yellow.ToArgb()));
        resources.Add(CustomResource.CreateCustomResource(2, "Nancy Drewmore", Color.Green.ToArgb()));
        resources.Add(CustomResource.CreateCustomResource(3, "Pak Jang", Color.LightPink.ToArgb()));
        return resources;
    }

    static Random myRand = new Random();
    public static List<CustomAppointment> GetAppointments(List<CustomResource> resources) {
        List<CustomAppointment> appointments = new List<CustomAppointment>();
        foreach(CustomResource item in resources) {
            string subjPrefix = item.Name + "'s ";
            appointments.Add(CustomAppointment.CreateCustomAppointment(subjPrefix + "meeting", item.ResID, 2, 5, lastInsertedID++));
            appointments.Add(CustomAppointment.CreateCustomAppointment(subjPrefix + "travel", item.ResID, 3, 6, lastInsertedID++));
            appointments.Add(CustomAppointment.CreateCustomAppointment(subjPrefix + "phone call", item.ResID, 0, 10, lastInsertedID++));
        }
        return appointments;
    }


    public static SchedulerDataObject DataObject {
        get {
            SchedulerDataObject dataObject = new SchedulerDataObject();


            if(HttpContext.Current.Session["ResourcesList"] == null) {
                HttpContext.Current.Session["ResourcesList"] = GetResources();
            }
            dataObject.Resources = HttpContext.Current.Session["ResourcesList"] as List<CustomResource>;

            if(HttpContext.Current.Session["AppointmentsList"] == null) {
                HttpContext.Current.Session["AppointmentsList"] = GetAppointments(dataObject.Resources);
            }
            dataObject.Appointments = HttpContext.Current.Session["AppointmentsList"] as List<CustomAppointment>;
            return dataObject;
        }
    }

    static MVCxAppointmentStorage defaultAppointmentStorage;
    public static MVCxAppointmentStorage DefaultAppointmentStorage {
        get {
            if(defaultAppointmentStorage == null) {
                defaultAppointmentStorage = CreateDefaultAppointmentStorage();
            }
            return defaultAppointmentStorage;

        }
    }

    static MVCxAppointmentStorage CreateDefaultAppointmentStorage() {
        MVCxAppointmentStorage appointmentStorage = new MVCxAppointmentStorage();
        appointmentStorage.AutoRetrieveId = true;
        appointmentStorage.Mappings.AppointmentId = "ID";
        appointmentStorage.Mappings.Start = "StartTime";
        appointmentStorage.Mappings.End = "EndTime";
        appointmentStorage.Mappings.Subject = "Subject";
        appointmentStorage.Mappings.AllDay = "AllDay";
        appointmentStorage.Mappings.Description = "Description";
        appointmentStorage.Mappings.Label = "Label";
        appointmentStorage.Mappings.Location = "Location";
        appointmentStorage.Mappings.RecurrenceInfo = "RecurrenceInfo";
        appointmentStorage.Mappings.ReminderInfo = "ReminderInfo";
        appointmentStorage.Mappings.ResourceId = "OwnerId";
        appointmentStorage.Mappings.Status = "Status";
        appointmentStorage.Mappings.Type = "EventType";

        appointmentStorage.CustomFieldMappings.Add(new DevExpress.Web.ASPxScheduler.ASPxAppointmentCustomFieldMapping("AppointmentCustomField", "CustomInfo"));
        return appointmentStorage;
    }

    static MVCxResourceStorage defaultResourceStorage;
    public static MVCxResourceStorage DefaultResourceStorage {
        get {
            if(defaultResourceStorage == null) {
                defaultResourceStorage = CreateDefaultResourceStorage();
            }
            return defaultResourceStorage;

        }
    }   

    static MVCxResourceStorage CreateDefaultResourceStorage() {
        MVCxResourceStorage resourceStorage = new MVCxResourceStorage();
        resourceStorage.Mappings.ResourceId = "ResID";
        resourceStorage.Mappings.Caption = "Name";
        resourceStorage.Mappings.Color = "Color";
        return resourceStorage;
    }

    public static SchedulerSettings GetSchedulerSettings() {
        return GetSchedulerSettings(null);
    }

    public static SchedulerSettings GetSchedulerSettings(this System.Web.Mvc.HtmlHelper customHtml) {
        SchedulerSettings settings = new SchedulerSettings();
        settings.Name = "scheduler";
        settings.CallbackRouteValues = new { Controller = "Home", Action = "SchedulerPartial" };
        settings.EditAppointmentRouteValues = new { Controller = "Home", Action = "EditAppointment" };
        settings.CustomActionRouteValues = new { Controller = "Home", Action = "CustomCallBackAction" };

        settings.Storage.Appointments.Assign(SchedulerDataHelper.DefaultAppointmentStorage);
        settings.Storage.Resources.Assign(SchedulerDataHelper.DefaultResourceStorage);

        settings.Storage.EnableReminders = false;
        settings.GroupType = SchedulerGroupType.Resource;
        settings.Views.DayView.Styles.ScrollAreaHeight = 400;
        settings.Start = DateTime.Now;

        settings.AppointmentFormShowing = (sender, e) => {
            var scheduler = sender as MVCxScheduler;
            if(scheduler != null)
                e.Container = new CustomAppointmentTemplateContainer(scheduler);
        };

        settings.OptionsForms.SetAppointmentFormTemplateContent(c => {
            var container = (CustomAppointmentTemplateContainer)c;
            ModelAppointment modelAppointment = new ModelAppointment() {
                ID = container.Appointment.Id == null ? -1 : (int)container.Appointment.Id,
                Subject = container.Appointment.Subject,
                Location = container.Appointment.Location,
                StartTime = container.Appointment.Start,
                EndTime = container.Appointment.End,
                AllDay = container.Appointment.AllDay,
                Description = container.Appointment.Description,
                EventType = (int)container.Appointment.Type,
                Status = container.Appointment.StatusId,
                Label = container.Appointment.LabelId,
                CustomInfo = container.CustomInfo,
                OwnerId = Convert.ToInt32(container.Appointment.ResourceId)
            };

            customHtml.ViewBag.DeleteButtonEnabled = container.CanDeleteAppointment;

            (container.ResourceDataSource as ListEditItemCollection).RemoveAt(0);
            customHtml.ViewBag.ResourceDataSource = container.ResourceDataSource;
            customHtml.ViewBag.StatusDataSource = container.StatusDataSource;
            customHtml.ViewBag.LabelDataSource = container.LabelDataSource;
            //ViewBag.ReminderDataSource = container.ReminderDataSource;
            customHtml.RenderPartial("CustomAppointmentFormPartial", modelAppointment);
        });
        return settings;
    }

    static int lastInsertedID = 0;

    // CRUD operations implementation
    public static void InsertAppointments(CustomAppointment[] appts) {
        if(appts.Length == 0)
            return;

        List<CustomAppointment> appointmnets = HttpContext.Current.Session["AppointmentsList"] as List<CustomAppointment>;
        for(int i = 0; i < appts.Length; i++) {
            if(appts[i] != null) {
                appts[i].ID = lastInsertedID++;
                appointmnets.Add(appts[i]);
            }
        }
    }

    public static void UpdateAppointments(CustomAppointment[] appts) {
        if(appts.Length == 0)
            return;

        List<CustomAppointment> appointmnets = System.Web.HttpContext.Current.Session["AppointmentsList"] as List<CustomAppointment>;
        for(int i = 0; i < appts.Length; i++) {
            CustomAppointment sourceObject = appointmnets.First<CustomAppointment>(apt => apt.ID == appts[i].ID);
            appts[i].ID = sourceObject.ID;
            appointmnets.Remove(sourceObject);
            appointmnets.Add(appts[i]);
        }
    }

    public static void RemoveAppointments(CustomAppointment[] appts) {
        if(appts.Length == 0)
            return;

        List<CustomAppointment> appointmnets = HttpContext.Current.Session["AppointmentsList"] as List<CustomAppointment>;
        for(int i = 0; i < appts.Length; i++) {
            CustomAppointment sourceObject = appointmnets.First<CustomAppointment>(apt => apt.ID == appts[i].ID);
            appointmnets.Remove(sourceObject);
        }
    }
}